### ISSUE TITLE
Command Line Parsing : A command line ending with a single quote must end with whitespace, else wcslen hangs for last argument.

### ISSUE THREAD

**Issue by JVD66** [msys2/MSYS2-packages#2333]
## Describe the issue

Whenever I try to run a command in MSYS2 (or Cygwin, for that matter, also), where the command line
ends with a single quote character ( "'" : ASCII : \x27 ) ,  without that single quote being followed by some
white space ( eg. ' ' : ASCII : \x20 ) ,  then if that program is a Windows program that uses the windows 
CRT utility function to obtain its command line arguments : 
            LPWSTR *szArglist = CommandLineToArgvW(GetCommandLineW(), &nArgs);
, as it must if it is a Windows CRT application that provides
            void mainCRTStartup ( void ) ;
, the program hangs forever in 'CommandLineToArgvW' . 

This does not happen if double quotes are used, or if there is another argument OR whitespace after the final "'" 
character, which is balanced in the Commmand Line returned by 'GetCommandLineW' .

## Steps to Reproduce the Problem

1. For example, here is a program compiled with Visual Studio 2019 Community Edition's CL.EXE :

        #include <windows.h>
        #include <processenv.h>
        #include <memory.h>
        #include <shellapi.h>
        #include <stringapiset.h>
        #include <mmeapi.h>

        #include <malloc.h>
        #include <string.h>
        #include <stdbool.h>
        #include <stdio.h>
        #include <stdlib.h>

        typedef unsigned char byte_t, u8_t;
        typedef signed   char   i8_t;
        typedef unsigned short u16_t;
        typedef signed   short i16_t;
        typedef unsigned int   u32_t;
        typedef signed   int   i32_t;
        typedef unsigned long  UL_t;
        typedef signed   long  IL_t;
        #ifdef _LP64
        typedef unsigned long  u64_t;
        typedef signed   long  i64_t;
        #else
        typedef unsigned long  long u64_t;
        typedef signed   long  long i64_t;
        #endif

        bool utf8str
        ( register const u16_t  *ws  // wide string input
        , register const u16_t   nw  // number of 16-bit chars in wide string
        , register u8_t         **s  // pointer to pointer to output byte string, will be allocated if = 0 or reallocated if too small.
        , register u16_t        *ns  // pointer to number of output bytes, can be changed if too small.
        )
        { if ( (ws==((void*)0UL))
             ||(nw==0)
             ||(s==((u8_t**)0UL))
             ||(ns==((u16_t*)0UL))
             ) return false;
          if ( (*s == NULL) || (*ns < ((nw << 1)+1)))
          {  if (*s)
             {
               *s = realloc( *s, *ns = ((nw << 1)+1));
               // fprintf(stderr,"Realloc: %p %hu\n", *s, *ns);
             }else
             { *s = calloc(   1, *ns = ((nw << 1)+1));
               // fprintf(stderr,"Calloc: %p %hu\n", *s, *ns);
             }
          }//else
           // fprintf(stderr,"OK: %p %hu\n", *s, *ns);
          register int
            nmb  = WideCharToMultiByte
                   ( CP_UTF8                              // UINT   codePage
                   , WC_ERR_INVALID_CHARS                 // DWORD  dwFlags
                   , ws                                   // LPCWCH lpWideCharStr
                   , ((int)nw)                            // int    cchWideChar
                   , *s                                   // LPSTR  lpMultiByteStr
                   , ((int)*ns)                                  // int    cbMultiByte
                   , '\0'                                 // LPCCH  lpDefaultChar
                   , false                                // LPBOOL lpUsedDefaultChar
                   );
          if( (nmb > 0) && (nmb <= *ns) )
          { *ns = nmb;
            (*s)[*ns]='\0';
            return true;
          }
          return false;
        }

        inline
        void* memcpy16( register u8_t *dst, register const u8_t *src, register u16_t len )
        { register const
            u8_t *se = src + len ;
          for
          (; src != se
           ; ++src, ++dst
          ) *dst = *src;
          return (dst - len);
        }

        int main( int argc, const char *const* argv, const char *const*  envp )
        {
          const char
              *const* ap = argv
            , *const* ae = argv + (((argc > 0) && (argc < 65536)) ? argc : 0)
            ; // this doubles as a universal C runtime argument parsing tester:
          for (; ( ap != ae) && *ap; ++ap)
            printf("%s\n", *ap);
          printf("N MIDI Devs: %u ; N MIXER DEVS: %u \n", midiOutGetNumDevs(), mixerGetNumDevs());
          return 0;
        }

        void
        mainCRTStartup ( void )
        {  static
            u8_t buf [ 65535 ];
           LPWSTR *szArglist;
           LPWSTR  cmdLine = GetCommandLineW();
           u32_t   nArgs;
           char *bp ;
           u16_t bufsz;
           if( cmdLine != NULL )
           {
             szArglist = CommandLineToArgvW(cmdLine, &nArgs);
             fprintf(stderr,"ARGS: %ls %d %p\n", ((const u16_t*)cmdLine), nArgs, ((void*)szArglist));
             if(NULL == szArglist)
             {
               fprintf(stderr, "CommandLineToArgvW failed\n");
             } else
             if ( ( nArgs > 0 ) && (nArgs < 65535) )
             { fprintf(stderr, "N args: %u\n", nArgs);
               register i8_t
                 **argv = ((i8_t**) alloca( sizeof( i8_t * ) * nArgs ));
               register char
                 **ap = argv
               , **ae = argv + nArgs;
               register const char
                 *const* cp = ((const char *const*)szArglist)
               , *const* ce = ((const char *const*)szArglist) + nArgs
               ;
               bp = buf;
               for (; (cp < ce) && (ap < ae); ++ap, ++cp)
               {  bufsz = sizeof(buf) & 65535U;
                  fprintf(stderr, "%p (%p) %p\n", ((void*)cp), ((void*)*cp), ((void*)ap) );
                  if ( (*cp) && utf8str( ((const u16_t *const)*cp) , wcslen(((const u16_t*)*cp)), &bp, &bufsz) && (bufsz > 0))
                  { *ap = alloca( bufsz + 1 );
                    memcpy16(*ap, buf, bufsz);
                    *((*ap)+bufsz) = '\0';
                    // fprintf(stderr, "ARG: %p %s\n", *ap, *ap);
                  }else
                  if  (*cp && (cp < (ce -1)))
                  { // fprintf(stderr, "bad arg %hu\n", ((u16_t)(ap-argv)));
                    *ap = NULL;
                  }else
                  { //fprintf(stderr, "bad arg %p\n", *cp);
                    *ap = NULL;
                  }
                  fprintf(stderr, "%p < %p : %c , %p < %p : %c\n", (void*)cp, (void*)ce, (cp < ce) ? '1' : '0', (void*)ap, (void*) ae, (ap < ae) ? '1' : '0');
               }
               //fprintf(stderr, "%p %p , %p %p\n", (void*)cp, (void*)ce, (void*)ap, (void*) ae);
               //fprintf(stderr, "%d %p\n", nArgs, argv);
               register
               int rv = main( nArgs, argv,  (const char *const*)environ );
               register
               bool status = ( 0 == rv );
               // Free memory allocated for CommandLineToArgvW arguments.
               LocalFree(szArglist);
               if (! status )
                 fprintf(stderr,"non-zero return from main(): %hhu\n", ((unsigned char)(rv & 255)));
             }else
               fprintf(stderr, "No Args: %p %d\n", ((void*)szArglist), nArgs);
           }else
           { const char *const* argv[]={"($0)",NULL};
             register
             int rv = main(1, argv,  (const char *const*)environ);
             bool status = ( 0 == rv );
             if (! status )
               fprintf(stderr,"non-zero return from main(): %hhu\n", ((unsigned char)(rv & 255)));
           }
        }

  Normally, this program compiles & runs fine from the MSYS2 command line, if the MSVC 2019 
  directories are in $PATH, and $INCLUDE and $LIB are set correctly : 

         $  CL tMidiDevs.c /LINK ucrt.lib shell32.lib kernel32.lib winmm.lib
          
         $ ./tMidiDevs.exe 's 1' 's 2' 's 3' # there is a space after "'s 3' " .
         ARGS: C:\Users\JVD\src\tMidiDevs.exe "s 1" "s 2" "s 3" 4 0000016124E18AE0
         N args: 4
         0000016124E18AE0 (0000016124E18B08) 000000358EBBFA80
         0000016124E18AE0 < 0000016124E18B00 : 1 , 000000358EBBFA80 < 000000358EBBFAA0 : 1
         0000016124E18AE8 (0000016124E18B46) 000000358EBBFA88
         0000016124E18AE8 < 0000016124E18B00 : 1 , 000000358EBBFA88 < 000000358EBBFAA0 : 1
         0000016124E18AF0 (0000016124E18B4E) 000000358EBBFA90
         0000016124E18AF0 < 0000016124E18B00 : 1 , 000000358EBBFA90 < 000000358EBBFAA0 : 1
         0000016124E18AF8 (0000016124E18B56) 000000358EBBFA98
         0000016124E18AF8 < 0000016124E18B00 : 1 , 000000358EBBFA98 < 000000358EBBFAA0 : 1
         C:\Users\JVD\src\tMidiDevs.exe
         s 1
         s 2
         s 3
         N MIDI Devs: 1 ; N MIXER DEVS: 2

  But if I do not type a space or another argument after "'s 3'" , (no space after "'") :
  
          $ ./tMidiDevs.exe 's 1' 's 2' 's 3'
          ARGS: C:\Users\JVD\src\tMidiDevs.exe "s 1" "s 2" "s 3" 4 00000163E0F97CD0
          N args: 4
          00000163E0F97CD0 (00000163E0F97CF8) 0000000BE58FFB20
          00000163E0F97CD0 < 00000163E0F97CF0 : 1 , 0000000BE58FFB20 < 0000000BE58FFB40 : 1
          00000163E0F97CD8 (00000163E0F97D36) 0000000BE58FFB28
          00000163E0F97CD8 < 00000163E0F97CF0 : 1 , 0000000BE58FFB28 < 0000000BE58FFB40 : 1
          00000163E0F97CE0 (00000163E0F97D3E) 0000000BE58FFB30
          00000163E0F97CE0 < 00000163E0F97CF0 : 1 , 0000000BE58FFB30 < 0000000BE58FFB40 : 1
          00000163E0F97CE8 (00000163E0F97D46) 0000000BE58FFB38
          00000163E0F97CE8 < 00000163E0F97CF0 : 1 , 0000000BE58FFB38 < 0000000BE58FFB40 : 1

   Now, I must press <CTRL>+C because the program has hung in the call to  wcslen() of the last argument.

   Note, I do not think there is an error in my program's argument parsing logic, it is simply trying to convert
   the last argument from a Wide Character string into a UTF-8 string, which apparently I must do even though
   MSYS64's locale program says I am in a UTF-8 locale.

   Only when the last string ends with "'" and there is no whitespace on the end of the last string, 
   the wcslen() function hangs .

   This does not happen when I run the same program from a CMD.EXE command line. 

    Why is this ?  Is MSYS64's bash messing up the last string if it is for a UTF-16 expecting Windows program and
    does not end with whitespace ?

## Additional Context: Operating System, Screenshots

- OS: [e.g. Windows 10 Pro build 20H2 , x86_64, Visual Studio 2019 , bash version: 4.4.23(2)-release  
- '...'

**Comment by JVD66** [comment]
Aha!  The hanging problem on last non-whitespace suffixed argument actually occurs in :
        @line 147:       LocalFree(szArglist);
as the attached file proves:


[tMidiDevs.c.txt](https://github.com/msys2/MSYS2-packages/files/5953942/tMidiDevs.c.txt)


When run with no whitespace on the end of a double-quoted line, sometimes it fails:

JVD@JVDSWIN MSYS ~/src
$ ./tMidiDevs.exe "s 1" "s 2" "s 3"
ARGS: C:\Users\JVD\src\tMidiDevs.exe "s 1" "s 2" "s 3" 4 0000013575440750
N args: 4
0000013575440750 (0000013575440778) 00000003C8F8FA50
ARG: 00000003C8F8FA30 C:\Users\JVD\src\tMidiDevs.exe
0000013575440750 < 0000013575440770 : 1 , 00000003C8F8FA50 < 00000003C8F8FA70 : 1
0000013575440758 (00000135754407B6) 00000003C8F8FA58
ARG: 00000003C8F8FA20 s 1
0000013575440758 < 0000013575440770 : 1 , 00000003C8F8FA58 < 00000003C8F8FA70 : 1
0000013575440760 (00000135754407BE) 00000003C8F8FA60
ARG: 00000003C8F8FA10 s 2
0000013575440760 < 0000013575440770 : 1 , 00000003C8F8FA60 < 00000003C8F8FA70 : 1
0000013575440768 (00000135754407C6) 00000003C8F8FA68
ARG: 00000003C8F8FA00 s 3
0000013575440768 < 0000013575440770 : 1 , 00000003C8F8FA68 < 00000003C8F8FA70 : 1
4 00000003C8F8FA50
Main Done: 1
C:\Users\JVD\src\tMidiDevs.exe
s 1
s 2
s 3
N MIDI Devs: 1 ; N MIXER DEVS: 2

JVD@JVDSWIN MSYS ~/src
$ ./tMidiDevs.exe "s 1" "s 2" "s 3"
ARGS: C:\Users\JVD\src\tMidiDevs.exe "s 1" "s 2" "s 3" 4 000002AD61003A80
N args: 4
000002AD61003A80 (000002AD61003AA8) 000000F9614FFC10
ARG: 000000F9614FFBF0 C:\Users\JVD\src\tMidiDevs.exe
000002AD61003A80 < 000002AD61003AA0 : 1 , 000000F9614FFC10 < 000000F9614FFC30 : 1
000002AD61003A88 (000002AD61003AE6) 000000F9614FFC18
ARG: 000000F9614FFBE0 s 1
000002AD61003A88 < 000002AD61003AA0 : 1 , 000000F9614FFC18 < 000000F9614FFC30 : 1
000002AD61003A90 (000002AD61003AEE) 000000F9614FFC20
ARG: 000000F9614FFBD0 s 2
000002AD61003A90 < 000002AD61003AA0 : 1 , 000000F9614FFC20 < 000000F9614FFC30 : 1
000002AD61003A98 (000002AD61003AF6) 000000F9614FFC28
ARG: 000000F9614FFBC0 s 3
000002AD61003A98 < 000002AD61003AA0 : 1 , 000000F9614FFC28 < 000000F9614FFC30 : 1
4 000000F9614FFC10
Main Done: 1
# I have to press CTRL+C 
So the program just hangs in the 'Main Done' state, without having printed arguments or actually running main()
(which would have printed something!), presumably because it detected that one of those arguments was bad,
and hangs on entry to LocalFree(szArglist) .  So, maybe I should not free that object, contrary to MS documentation.

Something very strange can happens with quoted arguments in this scenario. I'd love to hear any ideas why / 
how to avoid the hang / invalid arguments not allowing main() to even run .

**Comment by JVD66** [comment]
I'd really like to understand how to reliably call windows CRT main() type programs with any kind of QUOTED arguments reliably ; it doesn't seem to be possible with Windows command line parsing tools .

Best to write one's own multi-byte string token parser that handles quotes properly, and use it instead of any windows 
GetCommandLineArgvW ,  since this function seems to mess up memory somehow in this scenario, I guess.

**Comment by JVD66** [comment]
It even happens with this updated version of the program, using double quotes :
[tMidiDevs.c.txt](https://github.com/msys2/MSYS2-packages/files/5954080/tMidiDevs.c.txt)

The change was on line 59:
              , ((int)*ns) - (*ns ? 1 : 0)           // int    cbMultiByte
since we add one to *ns, bringing it out of bounds if 65535 , but this is highly unlikely to 
have caused the problem, which still sometimes happens with quoted arguments .

The same happens if a program is compiled with GCC / mingw64 / clang++ on Windows, they
all use a mainCRTStartup() very similar to mine and all suffer from the same problem.

Time to write a windows getdelim() implementation, I guess.

**Comment by JVD66** [comment]
Final updated version, still has problem:

[tMidiDevs.c.txt](https://github.com/msys2/MSYS2-packages/files/5954239/tMidiDevs.c.txt)

This version's output, when the problem occurs, shows NO OUTPUT AT ALL from the main() function, even
though this version of main() always prints something and is flushing its output, and apparently the call to main returns 0 :

$ ./tMidiDevs.exe "s 1" "s 2" 's 3'
ARGS: C 4 00000226821A3A80
N args: 4
00000226821A3A80 (00000226821A3AA8) 00000037E4FBF890
ARG: 00000037E4FBF870 [ 30 ] C:\Users\JVD\src\tMidiDevs.exe
00000226821A3A80 < 00000226821A3AA0 : 1 , 00000037E4FBF890 < 00000037E4FBF8B0 : 1
00000226821A3A88 (00000226821A3AE6) 00000037E4FBF898
ARG: 00000037E4FBF860 [ 3 ] s 1
00000226821A3A88 < 00000226821A3AA0 : 1 , 00000037E4FBF898 < 00000037E4FBF8B0 : 1
00000226821A3A90 (00000226821A3AEE) 00000037E4FBF8A0
ARG: 00000037E4FBF850 [ 3 ] s 2
00000226821A3A90 < 00000226821A3AA0 : 1 , 00000037E4FBF8A0 < 00000037E4FBF8B0 : 1
00000226821A3A98 (00000226821A3AF6) 00000037E4FBF8A8
ARG: 00000037E4FBF840 [ 3 ] s 3
00000226821A3A98 < 00000226821A3AA0 : 1 , 00000037E4FBF8A8 < 00000037E4FBF8B0 : 1
4 00000037E4FBF890
main(4): 00000037E4FBF890
C:\Users\JVD\src\tMidiDevs.exe
s 1
s 2
s 3
N MIDI Devs: 1 ; N MIXER DEVS: 2
Main Done: 1

JVD@JVDSWIN MSYS ~/src
$ ./tMidiDevs.exe "s 1" "s 2" 's 3'
ARGS: C 4 000001EC35113A80
N args: 4
000001EC35113A80 (000001EC35113AA8) 00000093C59FF860
ARG: 00000093C59FF840 [ 30 ] C:\Users\JVD\src\tMidiDevs.exe
000001EC35113A80 < 000001EC35113AA0 : 1 , 00000093C59FF860 < 00000093C59FF880 : 1
000001EC35113A88 (000001EC35113AE6) 00000093C59FF868
ARG: 00000093C59FF830 [ 3 ] s 1
000001EC35113A88 < 000001EC35113AA0 : 1 , 00000093C59FF868 < 00000093C59FF880 : 1
000001EC35113A90 (000001EC35113AEE) 00000093C59FF870
ARG: 00000093C59FF820 [ 3 ] s 2
000001EC35113A90 < 000001EC35113AA0 : 1 , 00000093C59FF870 < 00000093C59FF880 : 1
000001EC35113A98 (000001EC35113AF6) 00000093C59FF878
ARG: 00000093C59FF810 [ 3 ] s 3
000001EC35113A98 < 000001EC35113AA0 : 1 , 00000093C59FF878 < 00000093C59FF880 : 1
4 00000093C59FF860
main(4): 00000093C59FF860
C:\Users\JVD\src\tMidiDevs.exe
s 1
s 2
s 3
N MIDI Devs: 1 ; N MIXER DEVS: 2
Main Done: 1
# (this hangs, have to press CTRL+C)

**Comment by JVD66** [comment]
So what is most disconcerting is that the cancellation point of a "C" end-of-statement: ';' token ,
which ends a statement in which a function , main(), is ALWAYS called, which can only return 0 if it
prints and flushes a string, appears to have returned 0 and not printed any string ,
 (it cannot have been called at all!) , and the windows LocalFree() function hangs indefinitley, after
 all the string parsing has been done, the newly allocated last string on the stack was copied & printed OK, 
 the same problem happens if I use strdup() and malloc() instead of alloca() , it is not a problem with my
  string or memory allocation, it does appear to be a problem either with GetCommandLineW itself or with
  MSYS4's bash argument parsing and the vector it passes (somehow?) to what GetCommandLineW receives.

**Comment by JVD66** [comment]
That last version of main's first line is now:
  printf("main(%u): %p\n", argc, ((void*)argv));
before it even starts examining argv.  No output from this line is seen when the problem occurs, yet the caller of main() thinks main() returned 0 .
So main() was not called, because some windows bounds-checking mechanism has problems with my argv on the stack.
Can anyone please suggest why, or what that might be ? The addresses and sizes that program prints out show that my memory allocations & copies are working and do not overrun any bounds.

**Comment by JVD66** [comment]
ARG!  It was because I did not call exit() . Sorry!
Sometimes (not others - why), if the mainCRTStartup() does not explicitly call exit(), the process hangs forever.
Why this should happen sometimes and not others, and why much more frequently when the command line ends
with a quote with no whitespace after it, does not make any sense to me. 

Fixed version that does not have the problem calls exit() in mainCRTStartup : 

[tMidiDevs.c.txt](https://github.com/msys2/MSYS2-packages/files/5954438/tMidiDevs.c.txt)
